File: platforms/android/flow-contracts/src/test/kotlin/com/sloppad/android/flow/FlowRunner.kt
package com.sloppad.android.flow

data class FlowSnapshot(
    val activeTool: String,
    val session: String,
    val silent: Boolean,
    val sloppadCircle: String,
    val dotInnerIcon: String,
    val indicatorState: String,
    val bodyClass: String,
    val cursorClass: String,
)

class FlowRunner(
    private val platform: String,
    private val selectors: Map<String, String>,
) {
    private data class State(
        var tool: String = "pointer",
        var session: String = "none",
        var silent: Boolean = false,
        var circleExpanded: Boolean = false,
        var indicatorOverride: String = "",
    )

    private val state = State()

    fun run(flow: FlowDefinition) {
        reset(flow.preconditions)
        for (step in flow.steps) {
            if (shouldSkip(step)) {
                continue
            }
            runStep(step)
        }
    }

    private fun reset(preconditions: FlowPreconditions?) {
        state.tool = normalizeTool(preconditions?.tool ?: "pointer")
        state.session = normalizeSession(preconditions?.session ?: "none")
        state.silent = preconditions?.silent ?: false
        state.circleExpanded = false
        state.indicatorOverride = normalizeIndicator(preconditions?.indicatorState ?: "")
    }

    private fun shouldSkip(step: FlowStep): Boolean {
        val platforms = step.platforms ?: return false
        return platform !in platforms
    }

    private fun runStep(step: FlowStep) {
        when (step.action) {
            "tap" -> tap(requireTarget(step))
            "tap_outside" -> state.circleExpanded = false
            "verify" -> step.target?.let(::resolveSelector)
            "wait" -> step.durationMs ?: 0
            else -> error("unsupported action ${step.action}")
        }
        assertExpectations(step.expect)
    }

    private fun tap(target: String) {
        resolveSelector(target)
        when (target) {
            "sloppad_circle_dot" -> state.circleExpanded = !state.circleExpanded
            "sloppad_circle_segment_pointer" -> state.tool = "pointer"
            "sloppad_circle_segment_highlight" -> state.tool = "highlight"
            "sloppad_circle_segment_ink" -> state.tool = "ink"
            "sloppad_circle_segment_text_note" -> state.tool = "text_note"
            "sloppad_circle_segment_prompt" -> state.tool = "prompt"
            "sloppad_circle_segment_dialogue" -> toggleSession("dialogue")
            "sloppad_circle_segment_meeting" -> toggleSession("meeting")
            "sloppad_circle_segment_silent" -> state.silent = !state.silent
            "indicator_border" -> {
                state.session = "none"
                state.indicatorOverride = ""
            }
            else -> error("unsupported tap target $target")
        }
    }

    private fun toggleSession(next: String) {
        state.session = if (state.session == next) "none" else next
        state.indicatorOverride = ""
    }

    private fun requireTarget(step: FlowStep): String {
        return step.target?.takeIf { it.isNotBlank() }
            ?: error("missing target for action ${step.action}")
    }

    private fun resolveSelector(target: String): String {
        return selectors[target]?.takeIf { it.isNotBlank() }
            ?: error("missing selector for target $target")
    }

    private fun assertExpectations(expect: FlowExpectations?) {
        if (expect == null) {
            return
        }
        val snapshot = snapshot()
        expect.activeTool?.let { check(snapshot.activeTool == it) { "expected active_tool $it, got ${snapshot.activeTool}" } }
        expect.session?.let { check(snapshot.session == it) { "expected session $it, got ${snapshot.session}" } }
        expect.silent?.let { check(snapshot.silent == it) { "expected silent $it, got ${snapshot.silent}" } }
        expect.sloppadCircle?.let { check(snapshot.sloppadCircle == it) { "expected sloppad_circle $it, got ${snapshot.sloppadCircle}" } }
        expect.dotInnerIcon?.let { check(snapshot.dotInnerIcon == it) { "expected dot_inner_icon $it, got ${snapshot.dotInnerIcon}" } }
        expect.bodyClassContains?.let {
            check(snapshot.bodyClass.contains(it)) {
                "expected body_class to contain $it, got ${snapshot.bodyClass}"
            }
        }
        expect.indicatorState?.let {
            check(snapshot.indicatorState == it) { "expected indicator_state $it, got ${snapshot.indicatorState}" }
        }
        expect.cursorClass?.let { check(snapshot.cursorClass == it) { "expected cursor_class $it, got ${snapshot.cursorClass}" } }
    }

    private fun snapshot(): FlowSnapshot {
        val indicatorState = currentIndicatorState()
        val bodyClass = listOf(
            "tool-${state.tool}",
            "session-${state.session}",
            "indicator-$indicatorState",
            if (state.silent) "silent-on" else "silent-off",
            if (state.circleExpanded) "circle-expanded" else "circle-collapsed",
        ).joinToString(" ")
        return FlowSnapshot(
            activeTool = state.tool,
            session = state.session,
            silent = state.silent,
            sloppadCircle = if (state.circleExpanded) "expanded" else "collapsed",
            dotInnerIcon = toolIconId(state.tool),
            indicatorState = indicatorState,
            bodyClass = bodyClass,
            cursorClass = "tool-${state.tool}",
        )
    }

    private fun currentIndicatorState(): String {
        if (state.indicatorOverride.isNotBlank()) {
            return state.indicatorOverride
        }
        return when (state.session) {
            "dialogue" -> "listening"
            "meeting" -> "paused"
            else -> "idle"
        }
    }

    private fun normalizeTool(value: String): String {
        return when (value) {
            "pointer", "highlight", "ink", "text_note", "prompt" -> value
            else -> "pointer"
        }
    }

    private fun normalizeSession(value: String): String {
        return when (value) {
            "none", "dialogue", "meeting" -> value
            else -> "none"
        }
    }

    private fun normalizeIndicator(value: String): String {
        return when (value) {
            "idle", "listening", "paused", "recording", "working" -> value
            else -> ""
        }
    }

    private fun toolIconId(tool: String): String {
        return when (tool) {
            "highlight" -> "marker"
            "ink" -> "pen_nib"
            "text_note" -> "sticky_note"
            "prompt" -> "mic"
            else -> "arrow"
        }
    }
}

File: platforms/ios/Sources/SlopshellFlowContract/FlowRunner.swift
import Foundation

public struct FlowSnapshot {
    let activeTool: String
    let session: String
    let silent: Bool
    let slopshellCircle: String
    let dotInnerIcon: String
    let indicatorState: String
    let bodyClass: String
    let cursorClass: String
}

public enum FlowRunnerError: Error, CustomStringConvertible {
    case missingSelector(String)
    case unsupportedAction(String)
    case missingTarget(String)
    case expectationFailed(String)

    public var description: String {
        switch self {
        case .missingSelector(let target):
            return "missing selector for target \(target)"
        case .unsupportedAction(let action):
            return "unsupported action \(action)"
        case .missingTarget(let action):
            return "missing target for action \(action)"
        case .expectationFailed(let detail):
            return detail
        }
    }
}

public final class FlowRunner {
    private struct State {
        var tool = "pointer"
        var session = "none"
        var silent = false
        var circleExpanded = false
        var indicatorOverride = ""
    }

    private let platform: String
    private let selectors: [String: String]
    private var state = State()

    public init(platform: String, selectors: [String: String]) {
        self.platform = platform
        self.selectors = selectors
    }

    public func run(flow: FlowDefinition) throws {
        reset(using: flow.preconditions)
        for step in flow.steps {
            if shouldSkip(step: step) {
                continue
            }
            try run(step: step)
        }
    }

    private func reset(using preconditions: FlowPreconditions?) {
        state.tool = normalizeTool(preconditions?.tool ?? "pointer")
        state.session = normalizeSession(preconditions?.session ?? "none")
        state.silent = preconditions?.silent ?? false
        state.circleExpanded = false
        state.indicatorOverride = normalizeIndicator(preconditions?.indicatorState ?? "")
    }

    private func shouldSkip(step: FlowStep) -> Bool {
        guard let platforms = step.platforms else {
            return false
        }
        return platforms.contains(platform) == false
    }

    private func run(step: FlowStep) throws {
        switch step.action {
        case "tap":
            try tap(target: requireTarget(step))
        case "tap_outside":
            state.circleExpanded = false
        case "verify":
            if let target = step.target {
                _ = try resolveSelector(for: target)
            }
        case "wait":
            _ = step.durationMS ?? 0
        default:
            throw FlowRunnerError.unsupportedAction(step.action)
        }
        try assertExpectations(step.expect)
    }

    private func tap(target: String) throws {
        _ = try resolveSelector(for: target)
        switch target {
        case "slopshell_circle_dot":
            state.circleExpanded.toggle()
        case "slopshell_circle_segment_pointer":
            state.tool = "pointer"
        case "slopshell_circle_segment_highlight":
            state.tool = "highlight"
        case "slopshell_circle_segment_ink":
            state.tool = "ink"
        case "slopshell_circle_segment_text_note":
            state.tool = "text_note"
        case "slopshell_circle_segment_prompt":
            state.tool = "prompt"
        case "slopshell_circle_segment_dialogue":
            toggleSession("dialogue")
        case "slopshell_circle_segment_meeting":
            toggleSession("meeting")
        case "slopshell_circle_segment_silent":
            state.silent.toggle()
        case "indicator_border":
            state.session = "none"
            state.indicatorOverride = ""
        default:
            throw FlowRunnerError.unsupportedAction("tap:\(target)")
        }
    }

    private func toggleSession(_ next: String) {
        state.session = state.session == next ? "none" : next
        state.indicatorOverride = ""
    }

    private func resolveSelector(for target: String) throws -> String {
        guard let selector = selectors[target], selector.isEmpty == false else {
            throw FlowRunnerError.missingSelector(target)
        }
        return selector
    }

    private func requireTarget(_ step: FlowStep) throws -> String {
        guard let target = step.target, target.isEmpty == false else {
            throw FlowRunnerError.missingTarget(step.action)
        }
        return target
    }

    private func assertExpectations(_ expectations: FlowExpectations?) throws {
        guard let expectations else {
            return
        }
        let snapshot = snapshot()
        if let activeTool = expectations.activeTool, snapshot.activeTool != activeTool {
            throw FlowRunnerError.expectationFailed("expected active_tool \(activeTool), got \(snapshot.activeTool)")
        }
        if let session = expectations.session, snapshot.session != session {
            throw FlowRunnerError.expectationFailed("expected session \(session), got \(snapshot.session)")
        }
        if let silent = expectations.silent, snapshot.silent != silent {
            throw FlowRunnerError.expectationFailed("expected silent \(silent), got \(snapshot.silent)")
        }
        if let slopshellCircle = expectations.slopshellCircle, snapshot.slopshellCircle != slopshellCircle {
            throw FlowRunnerError.expectationFailed("expected slopshell_circle \(slopshellCircle), got \(snapshot.slopshellCircle)")
        }
        if let dotInnerIcon = expectations.dotInnerIcon, snapshot.dotInnerIcon != dotInnerIcon {
            throw FlowRunnerError.expectationFailed("expected dot_inner_icon \(dotInnerIcon), got \(snapshot.dotInnerIcon)")
        }
        if let bodyClassContains = expectations.bodyClassContains, snapshot.bodyClass.contains(bodyClassContains) == false {
            throw FlowRunnerError.expectationFailed("expected body_class to contain \(bodyClassContains), got \(snapshot.bodyClass)")
        }
        if let indicatorState = expectations.indicatorState, snapshot.indicatorState != indicatorState {
            throw FlowRunnerError.expectationFailed("expected indicator_state \(indicatorState), got \(snapshot.indicatorState)")
        }
        if let cursorClass = expectations.cursorClass, snapshot.cursorClass != cursorClass {
            throw FlowRunnerError.expectationFailed("expected cursor_class \(cursorClass), got \(snapshot.cursorClass)")
        }
    }

    private func snapshot() -> FlowSnapshot {
        let indicatorState = currentIndicatorState()
        let bodyClass = [
            "tool-\(state.tool)",
            "session-\(state.session)",
            "indicator-\(indicatorState)",
            state.silent ? "silent-on" : "silent-off",
            state.circleExpanded ? "circle-expanded" : "circle-collapsed",
        ].joined(separator: " ")
        return FlowSnapshot(
            activeTool: state.tool,
            session: state.session,
            silent: state.silent,
            slopshellCircle: state.circleExpanded ? "expanded" : "collapsed",
            dotInnerIcon: toolIconID(for: state.tool),
            indicatorState: indicatorState,
            bodyClass: bodyClass,
            cursorClass: "tool-\(state.tool)"
        )
    }

    private func currentIndicatorState() -> String {
        if state.indicatorOverride.isEmpty == false {
            return state.indicatorOverride
        }
        switch state.session {
        case "dialogue":
            return "listening"
        case "meeting":
            return "paused"
        default:
            return "idle"
        }
    }

    private func normalizeTool(_ value: String) -> String {
        switch value {
        case "pointer", "highlight", "ink", "text_note", "prompt":
            return value
        default:
            return "pointer"
        }
    }

    private func normalizeSession(_ value: String) -> String {
        switch value {
        case "none", "dialogue", "meeting":
            return value
        default:
            return "none"
        }
    }

    private func normalizeIndicator(_ value: String) -> String {
        switch value {
        case "idle", "listening", "paused", "recording", "working":
            return value
        default:
            return ""
        }
    }

    private func toolIconID(for tool: String) -> String {
        switch tool {
        case "highlight":
            return "marker"
        case "ink":
            return "pen_nib"
        case "text_note":
            return "sticky_note"
        case "prompt":
            return "mic"
        default:
            return "arrow"
        }
    }
}
